File: eurostruct/engine/src/eurostruct_engine/schemas/organisation.py
"""Les formes de fil de l'entrée : fonder, inviter, administrer.

CE QU'AUCUN DE CES CORPS NE PORTE
----------------------------------
**L'identité de celui qui agit.** Elle vient du jeton, dérivée par
``project_backend_actor()`` dans PostgreSQL, et pas une seule de ces formes
n'a de champ pour elle. Deux formes portent l'identité de celui qui *subit* —
modifier l'adhésion d'un collègue — et la primitive refuse que ce soit
l'appelant.

CE QUE ``InvitationEmise`` PORTE UNE SEULE FOIS
------------------------------------------------
Le secret du lien. Il est tiré par l'API, la base n'en connaît que
l'empreinte, et cette réponse est le **seul** endroit où il apparaîtra jamais.
Ni la liste des invitations, ni les journaux, ni un rechargement de l'écran ne
le montreront de nouveau : il n'existe plus nulle part.

CE QU'AUCUNE DE CES RÉPONSES NE DIT
------------------------------------
« Habilité ». Un ``validating_engineer`` créé ici porte un rôle
d'**organisation** : il décide qui, dans ce bureau, atteste un livrable. Il ne
porte aucune habilitation **normative** — celle-là se prend par le quatre-yeux,
et le registre national reste à 0/29.
"""

from __future__ import annotations

from pydantic import Field

from .common import CountryCode, Strict

__all__ = [
    "AdhesionModifiee",
    "Invitation",
    "InvitationAcceptee",
    "InvitationCreation",
    "InvitationEmise",
    "JetonInvitation",
    "ListeInvitations",
    "ListeMembres",
    "Membre",
    "MembreModification",
    "Organisation",
    "OrganisationCreation",
]

#: LES CINQ RÔLES D'ORGANISATION, tels que l'énumération PostgreSQL les nomme.
#: Le contrat les répète pour que l'interface les propose sans les deviner ;
#: la frontière, elle, reste dans la base — un rôle inconnu y est refusé par
#: le type, avant toute politique.
ROLES = ("owner", "admin", "engineer", "validating_engineer", "viewer")


class OrganisationCreation(Strict):
    """Ce qu'une personne saisit pour fonder son bureau.

    AUCUN CHAMP NE DÉSIGNE LE FONDATEUR. C'est l'appelant, dérivé du jeton :
    fonder au nom de quelqu'un d'autre n'a pas de sens, et l'accepter dans le
    corps ferait de l'appartenance une simple affirmation.
    """

    name: str = Field(min_length=1, max_length=200)
    country: CountryCode
    display_name: str | None = Field(
        default=None, max_length=200,
        description="Le nom professionnel du fondateur dans ce bureau. Il "
                    "figurera sur les attestations qu'il signera ; sans lui, "
                    "la primitive d'attestation refuse.")
    professional_id: str | None = Field(
        default=None, max_length=100,
        description="Numéro d'inscription à l'ordre ou à la chambre "
                    "professionnelle. Il n'est vérifié par personne ici, et "
                    "aucune valeur n'est inventée : il est reproduit tel quel.")


class Organisation(Strict):
    """Un bureau, tel que l'écran d'entrée le montre."""

    organization_id: str
    name: str
    country: CountryCode
    member_role: str = Field(
        description="Le rôle de l'appelant DANS ce bureau. Dérivé de "
                    "l'adhésion, jamais déclaré.")


class InvitationCreation(Strict):
    """Ce qu'un owner ou un admin saisit pour accueillir quelqu'un.

    AUCUNE ADRESSE ÉLECTRONIQUE. Une invitation liée à une adresse ouvre
    l'énumération des comptes : « invitez untel@exemple.fr » répondrait
    différemment selon que le compte existe ou non, et l'on apprendrait qui
    travaille où. ``label`` est un aide-mémoire libre pour l'émetteur ; il
    n'entre dans aucune décision.
    """

    role: str = Field(
        description="Le rôle que l'invité aura. Un « admin » ne peut pas "
                    "inviter un « owner » : il donnerait plus que son propre "
                    "pouvoir.")
    label: str | None = Field(
        default=None, max_length=200,
        description="Aide-mémoire libre, pour que l'émetteur s'y retrouve. "
                    "N'entre dans aucune décision.")
    display_name: str | None = Field(
        default=None, max_length=200,
        description="Le nom professionnel sous lequel l'invité signera. POSÉ "
                    "PAR L'ORGANISATION, jamais par l'invité : quelqu'un qui "
                    "choisirait lui-même ce nom pourrait attester sous celui "
                    "d'un autre.")
    professional_id: str | None = Field(default=None, max_length=100)
    validity_days: int = Field(
        default=14, ge=1, le=90,
        description="Durée de validité du lien. Un lien qui n'expire jamais "
                    "est un mot de passe permanent.")


class InvitationEmise(Strict):
    """La réponse à une émission — et le SEUL endroit où le secret apparaît.

    ``token`` N'EST PAS EN BASE. PostgreSQL n'en détient que le sha256 : une
    fuite de sauvegarde, un journal trop bavard ou une lecture accidentelle ne
    rendent aucun lien utilisable. En contrepartie, ce secret ne peut pas être
    réaffiché : il faut le copier maintenant, ou révoquer et réémettre.
    """

    invitation_id: str
    organization_id: str
    role: str
    expires_at: str
    token: str = Field(
        description="Le secret du lien, en clair, UNE SEULE FOIS. Il n'existe "
                    "nulle part ailleurs — ni en base, ni dans les journaux.")


class JetonInvitation(Strict):
    """Ce qu'un invité présente pour rejoindre un bureau.

    LE REFUS EST LE MÊME dans les quatre cas — inconnue, expirée, révoquée,
    déjà consommée. Distinguer « ce lien n'existe pas » de « ce lien a
    expiré » apprendrait à qui essaie des liens au hasard quand il a visé
    juste.
    """

    token: str = Field(min_length=1, max_length=512)


class InvitationAcceptee(Strict):
    """Ce que l'invité obtient : une organisation, et son rôle dedans."""

    organization_id: str
    organization_name: str
    member_role: str


class Invitation(Strict):
    """Une invitation, telle que le panneau d'administration la montre.

    NI LE SECRET NI SON EMPREINTE. Le premier n'existe plus ; la seconde
    suffirait à reconnaître un lien intercepté ailleurs, et n'aide en rien
    l'écran.
    """

    invitation_id: str
    role: str
    label: str | None
    display_name: str | None
    professional_id: str | None
    created_at: str
    expires_at: str
    accepted_at: str | None
    revoked_at: str | None
    state: str = Field(
        description="pending | accepted | revoked | expired. Calculé par la "
                    "base, pas par l'écran : deux horloges donneraient deux "
                    "réponses.")


class ListeInvitations(Strict):
    invitations: list[Invitation]


class Membre(Strict):
    """Une adhésion, telle que le panneau d'administration la montre.

    ``is_active = false`` NE FAIT PAS DISPARAÎTRE LA LIGNE, et c'est voulu
    depuis 0009 : une note de dix ans doit rester lisible et nommer son
    signataire. Ce qui disparaît, c'est l'accès.
    """

    user_id: str
    role: str
    display_name: str | None
    professional_id: str | None
    is_active: bool
    created_at: str
    deactivated_at: str | None
    is_me: bool = Field(
        description="Vrai pour la ligne de l'appelant. L'écran s'en sert pour "
                    "ne pas proposer des gestes que la base refuse de toute "
                    "façon : on ne modifie pas sa propre adhésion.")


class ListeMembres(Strict):
    members: list[Membre]


class MembreModification(Strict):
    """Ce qu'un owner ou un admin change sur l'adhésion d'un collègue.

    LES CHAMPS ABSENTS NE SONT PAS TOUCHÉS. Envoyer ``role`` seul ne remet pas
    les noms à zéro : ``update_names`` doit être demandé explicitement, faute
    de quoi un formulaire partiel effacerait le nom sous lequel quelqu'un a
    signé.
    """

    role: str | None = Field(default=None)
    is_active: bool | None = Field(default=None)
    display_name: str | None = Field(default=None, max_length=200)
    professional_id: str | None = Field(default=None, max_length=100)
    update_names: bool = Field(
        default=False,
        description="Sans lui, ``display_name`` et ``professional_id`` sont "
                    "ignorés. Un formulaire partiel n'efface pas le nom sous "
                    "lequel quelqu'un a signé.")


class AdhesionModifiee(Strict):
    """La confirmation d'une modification, relue depuis la base."""

    user_id: str
    role: str
    is_active: bool

File: eurostruct/web/lib/organisation.ts
/**
 * L'entrée dans l'application, vue du navigateur.
 *
 * CE QUE CET ÉCRAN REMPLACE
 * --------------------------
 * Rien. Il n'y avait rien. Un compte tout neuf, parfaitement authentifié,
 * arrivait devant une liste de projets vide — pas une erreur, pas une
 * explication, un écran nu — et aucun bouton ne permettait d'en sortir. La
 * seule façon d'exister dans l'application était un `insert` fait à la main
 * par le propriétaire de la base.
 *
 * AUCUNE FORME N'EST REDÉFINIE ICI
 * ---------------------------------
 * `Organisation`, `InvitationEmise`, `Membre` et les autres viennent du
 * contrat **généré** depuis les modèles Pydantic. L'une d'elles porte, une
 * seule fois, le secret d'une invitation : la recopier à la main quelque part
 * serait exactement la façon de le faire fuir dans un champ qu'on croyait
 * anodin.
 *
 * LE SECRET N'EST NI STOCKÉ NI RÉAFFICHABLE
 * -------------------------------------------
 * Il arrive dans la réponse de `emettreInvitation`, et nulle part ailleurs. Le
 * navigateur ne l'écrit pas en `localStorage` : un secret rangé dans le
 * navigateur d'une personne survivrait à sa session, et c'est exactement ce
 * qu'un lien à usage unique ne doit pas faire.
 *
 * AUCUN RÔLE N'EST JUGÉ ICI
 * --------------------------
 * « Un admin ne peut pas inviter un owner », « on ne modifie pas sa propre
 * adhésion », « le dernier propriétaire actif ne disparaît pas » : les trois
 * règles sont dans PostgreSQL. L'écran s'en sert pour **montrer ou
 * expliquer** ; la frontière n'est pas ici.
 */
import type {
  AdhesionModifiee,
  Invitation,
  InvitationAcceptee,
  InvitationCreation,
  InvitationEmise,
  ListeInvitations,
  ListeMembres,
  Membre,
  MembreModification,
  Organisation,
  OrganisationCreation,
} from "@contracts/generated/engine";
import { appelProtege, type PorteurDeJeton } from "@/lib/transport";

export type {
  AdhesionModifiee,
  Invitation,
  InvitationAcceptee,
  InvitationCreation,
  InvitationEmise,
  ListeInvitations,
  ListeMembres,
  Membre,
  MembreModification,
  Organisation,
  OrganisationCreation,
};

/** Les cinq rôles d'organisation, dans l'ordre décroissant de pouvoir. */
export const ROLES = [
  "owner",
  "admin",
  "engineer",
  "validating_engineer",
  "viewer",
] as const;

/** Ce que chaque rôle fait, en une phrase, pour l'écran d'invitation. */
export const ROLE_EXPLIQUE: Record<string, string> = {
  owner: "administre le bureau, et ne peut pas disparaître s'il est le dernier",
  admin: "administre les membres, sans pouvoir créer de propriétaire",
  engineer: "calcule et rédige les livrables",
  validating_engineer: "atteste et émet ; il ne rédige pas ce qu'il atteste",
  viewer: "lit, et rien d'autre",
};

/** Les deux rôles qui administrent. L'écran s'en sert pour MONTRER. */
export function peutAdministrer(organisation: Organisation | null): boolean {
  return organisation?.member_role === "owner"
    || organisation?.member_role === "admin";
}

/**
 * Les bureaux de l'appelant, et son rôle dans chacun.
 *
 * IDEMPOTENT: c'est une lecture. Elle peut être rejouée après un 401 sans rien
 * créer.
 *
 * SÉPARÉE DE `listerProjets` DÉLIBÉRÉMENT. « Zéro projet et zéro bureau » et
 * « zéro projet et un bureau » sont deux écrans différents — « créez votre
 * bureau » d'un côté, « créez votre premier projet » de l'autre — et une seule
 * liste vide ne permet pas de les distinguer.
 */
export async function listerOrganisations(
  porteur: PorteurDeJeton,
): Promise<Organisation[]> {
  const liste = await appelProtege<Organisation[]>(
    "/v1/organizations", porteur, { methode: "GET", idempotent: true },
  );
  return liste ?? [];
}

/**
 * Fonde un bureau et rend **le bureau**, pas son identifiant.
 *
 * PAS IDEMPOTENT AU SENS DU TRANSPORT — c'est une création — mais la primitive
 * l'est: un second appel avec le même nom rend le bureau déjà créé. Un
 * double-clic ne fonde donc pas deux bureaux, et l'écran n'a pas à s'en
 * préoccuper.
 */
export async function fonderOrganisation(
  porteur: PorteurDeJeton,
  brouillon: OrganisationCreation,
): Promise<Organisation> {
  const cree = await appelProtege<Organisation>(
    "/v1/organizations", porteur, { methode: "POST", corps: brouillon },
  );
  if (!cree?.organization_id) {
    // 201 SANS IDENTIFIANT N'EST PAS UNE FONDATION. On refuse plutôt que
    // d'afficher un bureau vide dont personne ne pourra rien faire.
    throw new Error("la fondation n'a rendu aucun bureau.");
  }
  return cree;
}

/**
 * Émet un lien d'invitation. **La réponse porte le secret, une seule fois.**
 *
 * Il n'existe nulle part ailleurs — ni en base, ni dans les journaux, ni dans
 * un rechargement de l'écran. L'appelant doit le montrer maintenant.
 */
export async function emettreInvitation(
  porteur: PorteurDeJeton,
  organizationId: string,
  brouillon: InvitationCreation,
): Promise<InvitationEmise> {
  const emise = await appelProtege<InvitationEmise>(
    `/v1/organizations/${encodeURIComponent(organizationId)}/invitations`,
    porteur, { methode: "POST", corps: brouillon },
  );
  if (!emise?.token) {
    throw new Error("l'emission n'a rendu aucun lien.");
  }
  return emise;
}

/** Les invitations du bureau — sans le secret ni son empreinte. */
export async function listerInvitations(
  porteur: PorteurDeJeton,
  organizationId: string,
): Promise<Invitation[]> {
  const liste = await appelProtege<ListeInvitations>(
    `/v1/organizations/${encodeURIComponent(organizationId)}/invitations`,
    porteur, { methode: "GET", idempotent: true },
  );
  return liste?.invitations ?? [];
}

/** Révoque un lien encore en attente. Un lien déjà consommé ne se révoque pas. */
export async function revoquerInvitation(
  porteur: PorteurDeJeton,
  organizationId: string,
  invitationId: string,
): Promise<void> {
  await appelProtege<null>(
    `/v1/organizations/${encodeURIComponent(organizationId)}`
    + `/invitations/${encodeURIComponent(invitationId)}`,
    porteur, { methode: "DELETE" },
  );
}

/**
 * Rejoint un bureau avec un lien.
 *
 * LE SECRET NE PASSE PAS PAR L'URL. Une adresse voyage dans l'historique du
 * navigateur, dans les journaux d'un proxy et dans l'en-tête `Referer` de la
 * page suivante ; le corps d'un POST, non.
 */
export async function accepterInvitation(
  porteur: PorteurDeJeton,
  token: string,
): Promise<InvitationAcceptee> {
  const acceptee = await appelProtege<InvitationAcceptee>(
    "/v1/invitations/accept", porteur, { methode: "POST", corps: { token } },
  );
  if (!acceptee?.organization_id) {
    throw new Error("l'acceptation n'a rendu aucun bureau.");
  }
  return acceptee;
}

/** L'équipe du bureau, adhésions désactivées comprises. */
export async function listerMembres(
  porteur: PorteurDeJeton,
  organizationId: string,
): Promise<Membre[]> {
  const liste = await appelProtege<ListeMembres>(
    `/v1/organizations/${encodeURIComponent(organizationId)}/members`,
    porteur, { methode: "GET", idempotent: true },
  );
  return liste?.members ?? [];
}

/**
 * Change le rôle, l'état ou le nom professionnel d'un collègue.
 *
 * `update_names` DOIT ÊTRE DEMANDÉ. Sans lui, `display_name` et
 * `professional_id` sont ignorés côté serveur : un formulaire partiel
 * n'efface pas le nom sous lequel quelqu'un a signé.
 */
export async function modifierMembre(
  porteur: PorteurDeJeton,
  organizationId: string,
  userId: string,
  modification: MembreModification,
): Promise<AdhesionModifiee> {
  const modifiee = await appelProtege<AdhesionModifiee>(
    `/v1/organizations/${encodeURIComponent(organizationId)}`
    + `/members/${encodeURIComponent(userId)}`,
    porteur, { methode: "PATCH", corps: modification },
  );
  if (!modifiee?.user_id) {
    throw new Error("la modification n'a rendu aucune adhesion.");
  }
  return modifiee;
}
